crates/rpc_client/src/lib.rs

```diff
@@ -0,0 +1,314 @@
+    /// or NULL value if included, If it is not included it is assumed to be a notification.
+    /// The value SHOULD normally not be Null and Numbers SHOULD NOT contain fractional parts.
+    ///
+    /// The Server **MUST** reply with the same value in the Response object if included.
+    /// This member is used to correlate the context between the two objects.
+    #[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash, Serialize, Deserialize)]
+    #[serde(deny_unknown_fields)]
+    #[serde(untagged)]
+    pub enum Id {
+        /// Numeric id
+        Num(u64),
+        /// String id
+        Str(String),
+    }
+    /// Represents JSON-RPC protocol version.
+    #[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+    pub enum Version {
+        /// Represents JSON-RPC 2.0 version.
+        V2_0,
+    }
+
+    impl Serialize for Version {
+        fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+        where
+            S: serde::Serializer,
+        {
+            match self {
+                Version::V2_0 => serializer.serialize_str("2.0"),
+            }
+        }
+    }
+
+    impl<'a> Deserialize<'a> for Version {
+        fn deserialize<D>(deserializer: D) -> Result<Version, D::Error>
+        where
+            D: serde::Deserializer<'a>,
+        {
+            deserializer.deserialize_identifier(VersionVisitor)
+        }
+    }
+
+    struct VersionVisitor;
+    impl<'a> serde::de::Visitor<'a> for VersionVisitor {
+        type Value = Version;
+
+        fn expecting(&self, formatter: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+            formatter.write_str("a string")
+        }
+
+        fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
+        where
+            E: serde::de::Error,
+        {
+            match value {
+                "2.0" => Ok(Version::V2_0),
+                _ => Err(serde::de::Error::custom(
+                    "Invalid JSON-RPC protocol version",
+                )),
+            }
+        }
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum GetTxByHashError {
+    #[error("Failed to send request")]
+    SendError { msg: String },
+
+    #[error("Failed to get response body")]
+    ResponseError { msg: String },
+
+    #[error("Response was not a Success<Transaction>")]
+    InterpretationError { msg: String, response_text: String },
+
+    #[error(transparent)]
+    OtherError(#[from] std::io::Error),
+}
+
+pub struct RpcClient {
+    url: String,
+    client: reqwest::blocking::Client,
+}
+
+impl RpcClient {
+    pub fn new(url: &str) -> Self {
+        RpcClient {
+            url: url.to_string(),
+            client: reqwest::blocking::Client::new(),
+        }
+    }
+
+    fn make_id() -> Result<u64, std::time::SystemTimeError> {
+        use std::time::{SystemTime, UNIX_EPOCH};
+        let since_epoch = SystemTime::now().duration_since(UNIX_EPOCH)?;
```

Comment:
Look into https://doc.rust-lang.org/std/sync/atomic/ for incrementing an ID while the function can maintain `&self`

Reply:
@feuGeneA did you try this approach?

Reply:
Addressed in https://github.com/NomicFoundation/hardhat/pull/3383/commits/9e60ef672a571be81f1dab27b28c60e182a162cd